satellite/gracefulexit/endpoint_test.go

```diff
@@ -0,0 +1,160 @@
+
+		c, err := client.Process(ctx, grpc.EmptyCallOption{})
+		require.NoError(t, err)
+
+		response, err := c.Recv()
+		require.NoError(t, err)
+
+		// should get a NotReady since the metainfo loop would not be finished at this point.
+		switch response.GetMessage().(type) {
+		case *pb.SatelliteMessage_NotReady:
+			// now check that the exiting node is initiated.
+			exitingNodeIDs, err := satellite.DB.OverlayCache().GetExitingNodes(ctx)
+			require.NoError(t, err)
+			require.Len(t, exitingNodeIDs, 1)
+
+			require.Equal(t, exitingNode.ID(), exitingNodeIDs[0])
+		default:
+			t.FailNow()
+		}
+		// close the old client
+		require.NoError(t, c.CloseSend())
+
+		// trigger the metainfo loop chore so we can get some pieces to transfer
+		satellite.GracefulExit.Chore.Loop.TriggerWait()
+
+		// make sure all the pieces are in the transfer queue
+		incompleteTransfers, err := satellite.DB.GracefulExit().GetIncomplete(ctx, exitingNode.ID(), numMessages+1, 0)
+		require.NoError(t, err)
+		require.Len(t, incompleteTransfers, numMessages)
+
+		// connect to satellite again to start receiving transfers
+		c, err = client.Process(ctx, grpc.EmptyCallOption{})
+		require.NoError(t, err)
+		defer func() {
+			err = errs.Combine(err, c.CloseSend())
+		}()
+
+		var pieceID storj.PieceID
+		for {
+			response, err = c.Recv()
+			if err == io.EOF {
+				// Done
+				break
+			}
+			require.NoError(t, err)
+
+			switch m := response.GetMessage().(type) {
+			case *pb.SatelliteMessage_TransferPiece:
+				require.NotNil(t, m)
+
+				// pick the first one to fail
+				if pieceID.IsZero() {
+					pieceID = m.TransferPiece.PieceId
+				}
+
+				if pieceID != m.TransferPiece.PieceId {
+					success := &pb.StorageNodeMessage{
+						Message: &pb.StorageNodeMessage_Succeeded{
+							Succeeded: &pb.TransferSucceeded{
+								OriginalPieceHash: &pb.PieceHash{PieceId: m.TransferPiece.PieceId},
+								AddressedOrderLimit: &pb.AddressedOrderLimit{
+									Limit: &pb.OrderLimit{
+										PieceId: m.TransferPiece.PieceId,
+									},
+								},
+							},
+						},
+					}
+					err = c.Send(success)
+					require.NoError(t, err)
+				} else {
+					failed := &pb.StorageNodeMessage{
+						Message: &pb.StorageNodeMessage_Failed{
+							Failed: &pb.TransferFailed{
+								PieceId: m.TransferPiece.PieceId,
+								Error:   pb.TransferFailed_UNKNOWN,
+							},
+						},
+					}
+					err = c.Send(failed)
+					require.NoError(t, err)
+				}
+			case *pb.SatelliteMessage_ExitCompleted:
+				// TODO test completed signature stuff
+				break
+			default:
+				t.FailNow()
+			}
+		}
+
+		// check that the exit has completed and we have the correct transferred/failed values
+		progress, err := satellite.DB.GracefulExit().GetProgress(ctx, exitingNode.ID())
+		require.NoError(t, err)
+
+		require.Equal(t, int64(numMessages), progress.PiecesTransferred)
+		// even though we failed 1, it eventually succeeded, so the count should be 0
```

Comment:
should we test for some pieces transferred that ultimately fail too or is the flow too similar?

Reply:
Yes.  I'll look into it.

Reply:
Updated